File: js/programmers/뒤에있는큰수찾기.js
function solution(numbers) {
    const answer = new Array(numbers.length).fill(-1);
    const stack = new Stack();
    stack.push(0);

    for (let i = 1; i < numbers.length; i++) {
      while (!stack.isEmpty() && numbers[stack.peek()] < numbers[i]) {
        answer[stack.pop()] = numbers[i];
      }
      stack.push(i);
    }

    return answer;
}

class Stack {
  constructor() {
    this.arr = [];
    this.index = -1;
  }

  isEmpty() {
    return this.index < 0;
  }

  push(item) {
    this.arr[++this.index] = item;
  }

  peek() {
    if (this.index < 0) {
      return null
    };
    return this.arr[this.index];
  }

  pop() {
    if (this.index < 0) {
      return null
    };
    return this.arr[this.index--];
  }
}

File: java/programmers/뒤에있는큰수찾기.java
import java.util.Arrays;
import java.util.Stack;
import java.util.stream.IntStream;

public class 뒤에있는큰수찾기 {
  public int[] solution(int[] numbers) {
    int[] answer = new int[numbers.length];
    Arrays.fill(answer, -1);
    Stack<Integer> stack = new Stack<Integer>();
    stack.push(0);

    for (int i : IntStream.range(1, numbers.length).toArray()) {
      while(!stack.isEmpty() && numbers[stack.peek()] < numbers[i]) {
        answer[stack.pop()] = numbers[i];
      }
      stack.push(i);
    }

    return answer;
  }
}
